docs/developer/source-versioning-2024-08-30.md

```diff
@@ -0,0 +1,183 @@
+# Design Document: Implementing Source Versioning / Tables from Sources in Terraform Provider for Materialize
+
+## 1. Introduction
+
+This design document outlines the implementation plan for supporting the new "Source Versioning / Tables from Sources" feature in the Terraform provider for Materialize.
+
+This feature aims to simplify the user experience, create a more unified model for managing data ingested from upstream systems, and provide more flexibility in handling upstream schema changes.
+
+## 2. Background
+
+Materialize is introducing changes to its source and table model:
+- The concept of a source will be unified around a single relation representing the progress of that source's ingestion pipeline.
+- Subsources will be replaced with tables created from sources.
+- Users will use a `CREATE TABLE .. FROM SOURCE ..` statement to ingest data from upstream systems.
+
+These changes require corresponding updates to the Terraform provider to maintain alignment with Materialize's data model and provide a smooth migration path for existing users.
+
+## 3. Objectives
+
+- Update the Terraform provider to support the new source and table model.
+- Provide a migration path for existing Terraform configurations.
+- Maintain backwards compatibility where possible.
+- Ensure the provider can work with both old and new versions of Materialize during the transition period.
+
+## 4. Design
+
+### 4.1 Schema Updates
+
+#### 4.1.1 Existing Source Resources
+
+We will maintain the existing source resources (e.g., `materialize_source_postgres`) but deprecate fields related to subsources:
+
+```go
+var sourcePostgresSchema = map[string]*schema.Schema{
+    // ... existing fields ...
+    "table": {
+        Description: "Tables to be ingested from the source. This field is deprecated and will be removed in a future version.",
+        Type:        schema.TypeSet,
+        Optional:    true,
+        Deprecated:  "Use the new `materialize_table_from_source` resource instead.",
+        Elem: &schema.Resource{
+            Schema: map[string]*schema.Schema{
+                // ... existing table schema ...
+            },
+        },
+    },
+}
+```
+
+#### 4.1.2 New Table From Source Resource
+
+Introduce a new `materialize_table_from_source` resource:
```

Comment:
I would call it `materialize_source_table`; the "table from source" is much more SQL-y than it is Terraform-y.

Reply:
Oh nice, I like this way better! Thanks for that suggestion!

Reply:
💯, that's also what we'll use for the dbt materialization (`source-table`).